### PR TITLE
T022 - Ligand-based screening: Neural networks

### PR DESCRIPTION
**Pull request by sakshimisra**
Talktorial ID: 022
* Title: Affinity prediction for EGFR
* Original authors: Ahmed Atta
* Reviewer(s): Sakshi Misra, Talia B. Kimber, Prof. Dr. Andrea Volkamer
* Date of review: 12/2020-present

# Content review

* Potential labels or categories (e.g. machine learning, small molecules, online APIs): artificial neural network, scatter plot, keras model.
* One line summary: Perform artificial neural network and predict the pIC50 values for unknown drug molecules.
* [x] The table of contents reflects the talktorial story-line; order of #, ##, ### headers is correct
* [x] URLs are linked with meaningful words, instead of pasting the URL directly or linking words like `here`.
* [x] I have spell-checked the notebook
* [x] Images have enough resolution to be rendered with quality, without being _too_ heavy.
* [x] All figures have a description
* [x] Markdown cell content is still in-line with code cell output (whenever results are discussed)
* [x] I have checked that cell outputs are not incredibly long (this applies also to `DataFrames`)
* [x] Formatting looks correctly on the Sphinx render (bold, italics, figure placing)

# Code review

* Time it took to execute (approx.):
* [x] Variable and function names follow snake case rules (e.g. `a_variable_name` vs `aVariableName`)
* [x] Spacing follows PEP8 (run Black on the code cells if needed)
* [x] Code line are under 99 characters each (run `black -l 99`)
* [x] Comments are useful and well placed
* [x] There are no unpythonic idioms like `for i in range(len(list))` (see slides)
* [x] All 3rd party dependencies are listed at the top of the notebook
* [x] I have marked all code cell with output referenced in markdown cells with the label `# TODO: CI`
* [x] I have identified potential candidates for a code refactor / useful functions
* [x] All `import ...` lines are at the top (practice part) cell, ordered by standard library / 3rd party packages / our own (`teachopencadd.*`)
* [x] I have update the relative paths to absolute paths.
* List here unfamiliar libraries you find in the imports and their intended use:

### CODE REVIEW

**Comment by dominiquesydow** [comment]
@sakshimisra  - your talktorial's final index will be `T022`. Could you please at some point update this PR according to that (let me know if you need help)?

**Comment by t-kimber** [comment]
Hi @sakshimisra !
The notebook is looking good. Here are a few comments to be addressed before asking for Andrea for final review:

- Please keep the structure of the template directory: https://github.com/volkamerlab/teachopencadd/tree/sm-021-review/teachopencadd/talktorials/000_template
   - README was deleted (it will be completed automatically, so no need to populate it). 
   -  all the generated or used data should live in the data folder
   - README in the data should be completed (by hand). 

- The sections "What is a neural network?", "Basic structure of a neural network", "What does a neuron do ?" are a bit redundant (and the figures as well: having one figure would suffice). Maybe you could have one section only which gives the general idea of the NN (mimics a human brain etc.) and then directly go into details with notations. 
- Double check your explanation for SGD and Mini-Batch Gradient Descent, it seems there might be confusion. 
- "Fault tolerance: They have potential for high fault tolerance. When these networks are scaled across multiple machines and multiple servers, they can debug and diagnose the problem by its own." I'm not too sure about this statement. Do you have a reference?
- Naming should be more explicit (e.g. "my_model" ?) 
- There are some `# NBVAL_CHECK_OUTPUT` missing when printing outputs. 
- The notebook is running fine up until `model = load_model("my_model")` but there are errors with this folder.

**Review by AndreaVolkamer — COMMENTED** [review]
@sakshimisra The notebook looks good, well done. I mainly included smaller textual adaptions. 
There are only a few general questions open that we can discuss after the presentation.

**Comment by sakshimisra** [comment]
Hi @AndreaVolkamer, thank you for your feedback :)

**Comment by jaimergp** [comment]
Thanks for the excellent work here @sakshimisra!
